### ISSUE TITLE
Loading referenced schemas from local file system using relative paths

### ISSUE THREAD

**Issue by shadow-cs** [gregsdennis/json-everything#8]
I'd like to be able to load references (`$ref` keyword) from local file system using relative paths.

I tried this as well with `Manatee.Json`. I figured out I have to set `DocumentPath` which concatenated the relative path with the absolute path I set to `DocumentPath` and then failed when separating the URI from the JSON Pointer to the schema within the file.

With `Newtonsoft.Json.Schema` this works by passing the following settings:

```cs
var schemaFullPath = Path.GetFullPath(schemaFile);
var settings = new JSchemaReaderSettings
{
  BaseUri = new Uri(schemaFullPath),
  Resolver = new JSchemaUrlResolver(),
};
```

To make this work each `JsonSchema` needs to have a `CurrentUri` (since the file locations on the file system may change, using `$id` isn't really an option) so that it can be set to `context.CurrentUri`.  This way `RefKeyword.Validate` will create a proper `Uri` pointing to the file to load. Dynamic loading of files that `Manatee` supports would be nice. Maybe a simple delegate that the user can set would do the job.

## Workaround

* Adding `$id` at runtime by mutating the JSON (parsing it, modifying and serializing) then passing it to `FromText` (which deserializes it again).
* Traversing the schema and registering all subschemas to `SchemaRegistry`.

This have the benefit of loading all schemas prior any validation.

However this isn't really a straight forward solution (as it requires about 200 LOC).

**Comment by gregsdennis** [comment]
Thanks for raising the issue.

I intentionally didn't support automatically loading schemas from references because this is the behavior set forth by the spec sections [8.2.2](http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.2) and [8.2.4](http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.4).   Specifically,

> (8.2.2, para 2) Note that this URI is an identifier and not necessarily a network locator. In the case of a network-addressable URL, a schema need not be downloadable from its canonical URI.

> (8.2.4, para 3) The resolved URI produced by these keywords is not necessarily a network locator, only an identifier. A schema need not be downloadable from the address if it is a network-addressable URL, and implementations SHOULD NOT assume they should perform a network operation when they encounter a network-addressable URI.

---

I'm happy to tag this for the future as an optional non-spec behavior, but I won't be adding it for now.  Instead, you'll need to explicitly load and register all the schemas you expect to use following [these instructions](https://gregsdennis.github.io/json-everything/usage/schema/references.html).

**Comment by shadow-cs** [comment]
Totally understand, but on the other hand loading schema from local file system is a bit different case in my view (since no network operation needs to be done). Also there usually is no `$id` (which make little sense for locally stored schemas).

Basically I'd like to see two improvements:

* Improve the base URI detection so it works better for locally stored files without an `$id`.
* Allow the schema to be more easily traversable, so we can load the referenced schemas more easily ourselves if we choose to do so. Or better yet provide an extensions to do so.

I'll be happy to send my current implementation of referenced schema loading if you're interested.

**Comment by gregsdennis** [comment]
Whether on disk, on a network, or in a database, automatic loading of schemas is explicitly unsupported.

The `JsonSchema.FromFile()` method is provided as a convenience.  My suggestion is to preload your schemas and register them in the global registry with the URIs that are being used to reference them.

> Allow the schema to be more easily traversable...

The library provides a very easily traversable object model.  I'm not sure what you're asking for here.

**Comment by shadow-cs** [comment]
> Whether on disk, on a network, or in a database, automatic loading of schemas is explicitly unsupported.

I'm not saying you should do this out-of-the-box, but providing an extension would certainly be nice. I'd even appreciate an option to provide `$id` (current document URI) externally so I didn't have to mutate the JSON.

> My suggestion is to preload your schemas and register them in the global registry with the URIs that are being used to reference them.

If you don't know all your schema files prior to parsing, this gets a little tricky. But I'm doing exactly this as you described in the docs and suggested above as well. See [gist](https://gist.github.com/shadow-cs/24576043ac49421910ae7cdbabf3e8b3).

> The library provides a very easily traversable object model. I'm not sure what you're asking for here.

Don't get me wrong, the schema is fairly easy traversable. But it could be even easier, see `JsonSchemaKeywordExtensions` in the gist. My problem with my solution is that I'm tied with current implementation of the keywords and any changes or addition of keywords might break the implementation.

If keywords that potentially have sub-schemas implemented some interface providing the enumerable, traversing would be even easier for me as a consumer.

**Comment by gregsdennis** [comment]
Regarding automatic loading, I could add a `Func<string, JsonSchema>` (take `$id`, return schema) property to `SchemaRegistry` that would do nothing by default.  You could then set this to a method that performs your fetch.  If a schema is returned it would be automatically added to the registry.

I did this with Manatee.Json, except that I had a default download method.  I don't want to have that as the default behavior here.

The keywords that contain subschemas do implement `IRefResolvable`.  This is used during subschema registration, a step that occurs prior to the main validation.  It also allows the reference search to drill down into the schema.  I didn't consider exposing the subschema(s) through the interface.  It would have to be a collection as some keywords can contain multiple subschemas.  That might conflate the purpose.  Or just a couple new interfaces.  I'm not sure.  I'll think about it.

What is your use case for manually traversing the schema?  My purpose was for `JsonSchema` to be somewhat black-boxed.

**Comment by shadow-cs** [comment]
> Regarding automatic loading, I could add a Func<string, JsonSchema> (take $id, return schema) property to SchemaRegistry that would do nothing by default. You could then set this to a method that performs your fetch. If a schema is returned it would be automatically added to the registry.
>
> I did this with Manatee.Json, except that I had a default download method. I don't want to have that as the default behavior here.

I totally get that, no need to provide it by default.

I know, I did experiment with `Manatee.Json` as well but I did have a hard time loading the locally stored referenced schemas (possibly due to a fact that we don't use `$id` in those locally stored schemas as the location on local file system is unknown, multiple people may checkout the code to any location through Git). When searching for solution I found this repo (in one of the issues) and tried to use it instead. With a success at the end.

I did have an issue with just a `string` (or `Uri`) in the delegate because it provides little context. Adding a schema that is being parsed/resolved would be nice or some other form how I can figure out `basePath` (current file path) and resolve the sub-schema reference (compute the correct full path from relative path). Again I may see this as a problem because `$id`s are not present in our schemas.

> The keywords that contain subschemas do implement IRefResolvable. This is used during subschema registration, a step that occurs prior to the main validation. It also allows the reference search to drill down into the schema. I didn't consider exposing the subschema(s) through the interface. It would have to be a collection as some keywords can contain multiple subschemas. That might conflate the purpose. Or just a couple new interfaces. I'm not sure.

Then I probably didn't understand the true purpose of `IRefResolvable` but just looking at it, it didn't fit mu purpose (during traversal). This is why exposing interfaces for sub-schema traversal made sense to me, having an interface with `IEnumerable<JsonSchema> GetSubSchemas()` would be a nice addition (see the gist linked above).

> I'll think about it.

Thanks.

> What is your use case for manually traversing the schema? My purpose was for JsonSchema to be somewhat black-boxed.

Once I load the entry-point schema, I don't know what other sub-schemas are referenced by any of the `$ref`s so I need to find them and register them to the registry (also see the gist for details, maybe you'll say, my approach is overly complex and there is a simpler solution, I just didn't see it).

**Comment by gregsdennis** [comment]
I started trying to add a subschemas property to the `IRefResolvable` interface, but I found that the main problem is that the keywords expose subschemas differently.  Some have a single subschema (`contains`), while others have multiple (`anyOf`).  For the ones with multiple, the subschema is a direct descendant in an array (`anyOf`), while for others, the schema is nested under keys (`$defs`, `properties`), both of which differently affect how references would be resolved.

This means that the "better" solution is actually to create three different interfaces (single schema, multiple indexed subschemas, multiple keyed subschemas) rather than adding onto `IRefResolvable`, but I'm not sure that really helps navigating the schema structure.
